### PR TITLE
HPCC-31981 Mutex code was inefficient

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: system/jlib/jmutex.hpp [hunk 1]
```diff
@@ -46,8 +47,9 @@ extern jlib_decl void spinUntilReady(std::atomic_uint &value);
 #endif
 
 #ifdef _WIN32
-class jlib_decl Mutex
+class jlib_decl OldMutex
 {
+friend class Monitor;
 protected:
     Mutex(const char *name)
```
Comment (ghalliday) [inline system/jlib/jmutex.hpp:54]: OldMutex - otherwise you get windows compile errors.

File: system/jlib/jmutex.hpp [hunk 2]
```diff
@@ -46,8 +47,9 @@ extern jlib_decl void spinUntilReady(std::atomic_uint &value);
 #endif
 
 #ifdef _WIN32
-class jlib_decl Mutex
+class jlib_decl OldMutex
```
Comment (ghalliday) [inline system/jlib/jmutex.hpp:50]: Legacy might be a better name than old, but doesn't really matter.

File: system/jlib/jpqueue.hpp [hunk 3]
```diff
@@ -102,13 +102,14 @@ template<typename T> class WaitQueue: public CInterface, protected Mutex
...
-        int locked = unlockAll();
+        mutex.unlock();
         ret=counter.wait(timeout);
-        lockAll(locked);
+        mutex.lock();
 
         waiting--;
         return ret;
     }
+    Mutex mutex;
```
Comment (ghalliday) [inline system/jlib/jpqueue.hpp:112]: trivial: newline before rather than after so that members are grouped together

File: system/jlib/jmutex.hpp [hunk 4]
```diff
@@ -165,23 +184,30 @@ class jlib_decl NamedMutex
...
+    TimedMutex threadmutex;
     char *mutexfname;
 };
 
+template<class T> class jlib_decl MutexBlock
+{
+private:
+    T &mutex;
+public:
+    MutexBlock<T>(T &m) : mutex(m) { mutex.lock(); };
```
Comment (ghalliday) [inline system/jlib/jmutex.hpp:196]: I think the <T> is not needed/correct on the constructor, and is causing compile errors for wasmembed

**Review by ghalliday — COMMENTED** [review]
@richardkchapman looks good.  Needs a bit of tidying up - commit message, squash and a few compile issues, but otherwise good to merge.

**Comment by richardkchapman** [comment]
@ghalliday made changes as suggested, and squashed. Please re-review

**Review by ghalliday — APPROVED** [review]

**Pull request merged**